In [2]:
#!wget --no-cache -O init.py -q https://raw.githubusercontent.com/rramosp/20201.ai4eng/master/init.py
import init; init.init(force_download=False); init.get_weblink()

endpoint https://m5knaekxo6.execute-api.us-west-2.amazonaws.com/dev-v0001/rlxmooc


In [186]:
from local.lib.rlxmoocapi import submit, session
student = session.Session(init.endpoint).login( course_id=init.course_id, 
                                                session_id="UDEA", 
                                                lab_id="L02.03" )
init.get_weblink()

username: kevin.martinez@udea.edu.co
password: ········
endpoint https://m5knaekxo6.execute-api.us-west-2.amazonaws.com/dev-v0001/rlxmooc


In [4]:
import numpy as np
import pandas as pd

# Task 1: Extract data

we have a dataframe of items with a price and, **sometimes**, with an extra column (`margin`).

You will have to complete a function that will **filter** the dataframe selecting the rows:

- whose price is > 100
- **OR** whose margin is >10, if the column `margin` is present

your function must return **A LIST** with the item ids of the selected rows

For instance, with the following dataframe

             price  category
    itemid                  
    56556   108.15       2.0
    73065    83.54       2.0
    36619   114.42       0.0
    73414    82.94       0.0
    13410   115.13       2.0
    66153    91.50       1.0
    77380    85.82       0.0
    73249    95.95       0.0
    11654   100.22       2.0
    11972    77.16       1.0

your must return the following list:

     [56556, 36619, 13410, 11654]
     
But if you get the following dataframe

             price  category  margin
    itemid                          
    39059    98.11       0.0   11.04
    19526    98.11       1.0   11.25
    78176    94.34       1.0   10.51
    50948   102.37       1.0   10.77
    12111    98.07       1.0    8.50
    56191    98.53       1.0   11.65
    38887    91.49       2.0   11.24
    77915   117.30       0.0    8.64
    55010    96.13       0.0    8.95
    45925    98.59       1.0   10.45
    
you must return the following list

    [39059, 19526, 78176, 50948, 56191, 38887, 77915, 45925]

In [5]:
def create_df(missing=False, n=10):
    itemid   = np.random.randint(100000, size=n)+1000
    category = np.random.randint(3, size=n)
    price    = np.round(np.random.normal(loc=100, scale=10, size=n),2)
    margin   = np.round(np.random.normal(loc=10, scale=1, size=n),2)
    
    if missing:
        nmissing = np.random.randint(len(price)//2)+2                                     
        price[np.random.permutation(len(price))[:nmissing]] = np.nan
    
    d = pd.DataFrame(np.r_[[price, category, margin]].T, index=itemid, columns=["price", "category", "margin"])
    d.index.name="itemid"
    if np.random.random()>.5:
        d = d[d.columns[:2]]
        
    return d

In [26]:
d = create_df()
d

price  category
itemid                  
22489    94.67       0.0
84440   101.96       2.0
73870   115.39       2.0
76548    91.37       2.0
80357    92.75       0.0
84162    86.38       2.0
95464   106.74       2.0
12811   100.92       1.0
93402   125.79       2.0
9382    111.91       2.0

In [29]:
def select_items(df):
    columns = df.columns
    if 'margin' in columns:
        result = df[(df.price > 100) | (df.margin > 10)].index.tolist()
    else:
        result = df[(df.price > 100)].index.tolist()
    return result

**manually check your answer**

In [30]:
select_items(d)

[84440, 73870, 95464, 12811, 93402, 9382]

**submit your code**

In [31]:
student.submit_task(globals(), task_id="task_01");

# Task 2: Group statistics

Complete the following function so that it returns a dataframe with the average, max and min **prices** per category. 

For instance, for the following dataframe

             price  category  margin
    itemid                          
    17946    93.85       1.0   10.64
    61190    91.72       1.0    9.76
    39639   100.16       1.0   10.67
    17791   110.44       2.0    9.65
    7333    101.05       1.0    9.69
    77362   122.33       0.0   11.14
    92646   108.13       2.0   10.58
    27797    85.52       2.0   10.88
    31746    97.56       0.0    9.75
    12355   101.04       2.0    9.51
    
you should return the following dataframe

                 media  maximo  minimo
    categoria                         
    0         109.9450  122.33  97.56
    1          96.6950  101.05  91.72
    2         101.2825  110.44  85.52
    
observe that your result
- must not be a **multilevel** columnset.
- the column names and the index name must be **exactly** as in the example.
- the **index** must be of type **int**.

In [105]:
def get_stats(df):
    
    result = pd.DataFrame(index = [0,1,2], columns = ['media', 'maximo', 'minimo'])
    result.index.name = 'categoria'
    result['media'] = df[['price', 'category']].groupby('category').mean()
    result['maximo'] = df[['price', 'category']].groupby('category').max()
    result['minimo'] = df[['price', 'category']].groupby('category').min()
    
    return result

**manually check your answer**

In [98]:
d = create_df()
d

price  category  margin
itemid                          
92720    73.48       1.0   10.92
19335    88.88       2.0   10.45
66262    88.97       1.0   10.03
93190   101.80       0.0   11.73
1576    119.77       2.0   10.16
48443    97.11       1.0   10.71
77653    79.34       2.0   10.20
22849   113.44       2.0    9.57
72235    89.88       1.0    9.71
13195   106.43       1.0   10.49

In [99]:
d.index.dtype

dtype('int64')

In [106]:
nf = get_stats(d)
nf

media  maximo  minimo
categoria                          
0          101.8000  101.80  101.80
1           91.1740  106.43   73.48
2          100.3575  119.77   79.34

In [107]:
nf.index.dtype

dtype('int64')

**submit your code**

In [108]:
student.submit_task(globals(), task_id="task_02");

# Task 3: Fill in missing data

Fill in the missing data in the **price** column with the following procedure:

- compute the mean and std of the avaialable prices
- sample from a normal distribution with the computed mean and std (see [`np.random.normal`])(https://numpy.org/doc/stable/reference/random/generated/numpy.random.normal.html), as many samples as missing data
- substitute the missing values with the samples

For instance, for this input dataframe:

             price  category  margin
    itemid                          
    18922      NaN       1.0   10.32
    69500   121.25       1.0   10.22
    76442    90.25       1.0   12.60
    33863   106.51       0.0   10.26
    15904    95.87       1.0   11.51
    41946   103.47       2.0    9.85
    85451    93.08       2.0    9.56
    70028   116.68       1.0    9.11
    26860      NaN       2.0    9.71
    12807    91.48       0.0    9.77
    
your solution might be similar to this (not exactly the same as you will be sampling data):


                 price  category  margin
    itemid                              
    18922    97.441188       1.0   10.32
    69500   121.250000       1.0   10.22
    76442    90.250000       1.0   12.60
    33863   106.510000       0.0   10.26
    15904    95.870000       1.0   11.51
    41946   103.470000       2.0    9.85
    85451    93.080000       2.0    9.56
    70028   116.680000       1.0    9.11
    26860   103.294843       2.0    9.71
    12807    91.480000       0.0    9.77
    
    
**WARN**: your function must not modify the original dataframe, make a copy of the input dataframe, fill the values in the copy and return it.

**HINT**: use the [`isna()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isna.html) method of a dataframe or a series.

In [221]:
def fillna(df):
    
    mean_std = df.describe()[['price']].loc['mean':'std'].to_numpy().tolist() # getting the mean and std
    
    na = df.isna()
    
    info = na[['price']].describe().to_numpy().tolist()
    if info[2][0] == False: 
        n_samples = info[0][0] - info[3][0]
    else: n_samples = info[3][0]
    
    new_samples = np.random.normal(mean_std[0][0], mean_std[1][0], n_samples)
    
    to_replace = na[na.price == True].index.tolist() # getting the rows to replace
    
    result = df.copy()
    
    i = 0 # replacing NaN values
    for index in to_replace:
        result.at[index, 'price'] = new_samples[i]
        i += 1
        
    return result

**check manually your code**

In [222]:
d = create_df(missing=True)
d

price  category
itemid                  
55860    98.39       1.0
5831     84.55       1.0
64358   104.73       0.0
33026    96.72       1.0
81521    80.60       2.0
82338   118.18       1.0
76433      NaN       0.0
2831    106.99       0.0
53022      NaN       2.0
8647    104.72       0.0

In [223]:
fillna(d)

price  category
itemid                      
55860    98.390000       1.0
5831     84.550000       1.0
64358   104.730000       0.0
33026    96.720000       1.0
81521    80.600000       2.0
82338   118.180000       1.0
76433   115.834559       0.0
2831    106.990000       0.0
53022   114.988148       2.0
8647    104.720000       0.0

**submit your code**

In [224]:
student.submit_task(globals(), task_id="task_03");